api/VsoClient.ts

```diff
@@ -183,41 +183,38 @@ export class VsoClient {
         return url.resolve(this.baseUrl, path.join(this.basePath, relativeUrl));
     }
 
-    private getSerializedObject(queryValue: any, object: any): string {
-        let value:string = "";
-        let first:boolean = true;
-
-        for (let property in object) {
-            if (object.hasOwnProperty(property)) {
-                let prop = object[property];
-                let valueString = this.getValueString(property, prop);
-                if (first && prop !== undefined) {
-                    value += valueString;
-                    first = false;
-                } else if (prop !== undefined) {
-                    value += "&" + valueString;
+    private queryParamsToStringHelper(queryParams: any, prefix: string): string {
+        if (queryParams == undefined) {
+            return '';
+        }
+        let queryString: string = '';
+
+        if(typeof(queryParams) != 'string'){
+            for (let property in queryParams) {
+                if (queryParams.hasOwnProperty(property)){
+                    let prop = queryParams[property];
+                    let newPrefix = prefix + encodeURIComponent(property.toString()) + '.';
+                    queryString += this.queryParamsToStringHelper(prop, newPrefix);
                 }
             }
         }
 
-        if (value == ""){
-            value += queryValue + "=" + object.toString();
+        if(queryString == '' && prefix.length > 0){
+            //Will always need to chop period off of end of prefix
+            queryString = prefix.slice(0,-1) + '=' + encodeURIComponent(queryParams.toString()) + '&';
         }
-        return value;
+        return queryString;
     }
 
-    private getValueString(queryValue, value) {
-        let valueString = null;
-        if (typeof(value) === 'object') {
-            valueString = this.getSerializedObject(queryValue, value);
-        } else {
-            valueString = queryValue + "=" + encodeURIComponent(value);
-        }
-        return valueString;
+    private queryParamsToString(queryParams: any): string {
+        let queryString: string = '?' + this.queryParamsToStringHelper(queryParams, '');
+
+        //Will always need to slice either a ? or & off of the end
+        return queryString.slice(0,-1);

```

Comment:
Why -1? How is this different than substring?

Reply:
This is equivalent to substring(0, queryString.length - 1). Slice and substring are very similar, I tend to like slice in cases like this due to clarity but that's just a preference. https://stackoverflow.com/questions/2243824/what-is-the-difference-between-string-slice-and-string-substring